# STAT 207 Lab Assignment 2 - Individual Part - [25 total points]

## Dataframe Manipulation and Cleaning

Due: Tuesday, January 30, 11:59pm CST

<hr>


## <u>Case Study</u>: Steam Game Play and Purchase Analysis
Steam is the world's most popular PC Gaming hub. With a massive collection that includes everything from AAA blockbusters to small indie titles, great discovery tools can be super valuable for Steam. We would like to use the following dataset below to learn more about Steam player behavior.

The dataset we will be analyzing is comprised of a random sample of 400 steam users and their game play and purchase behaviors. It has the following columns:
* **user_id** of the player
* **game_name**,
* **activity**:
    - <u>purchase</u>: indicating that the user has *purchased* the corresponding game
    - <u>play</u>: indicating that the user has *played* the corresponding game (for at least some amount of time.) 
* **hours_played_if_play**:
    - if the row corresponds to a 'play' activity, this number represents the number of hours the user has played the game
    - if the row corresponds to a 'purchase' activity, this number is always a 1 (and means nothing... it's a placeholder).

<hr>


### <u>Research Goals</u>

In this analysis, we will consider a game's **"all-star" player** to be the person who had played this game for the most amount of hours. With this definition in mind, we will do the following in this assignment.

1. **Missing Value Cleaning**: Identify any implicit or explicit missing values in the dataset and drop any rows that have a missing value.
2. **Game Purchase vs. Play Behaviors**: What percent of games that were *bought* were actually *played* in this dataset?
3. **Top Game Buyer**: Learn more about the person who bought the most amount of games in this dataset. How many of these games did they actually play?
4. **"All Star Players"**: Who were the "all-star players" of the games? How long did they play this game?
5. **All Star Player Behavior**: What percent of games that these "all star players" bought did they actually play?



### <u>Points </u>
<p style="margin-top:0in;margin-right:0in;margin-bottom:0in;margin-left:0in;font-size:11.0pt;font-family:&quot;Calibri&quot;,sans-serif;line-height:normal;"></p>
<table style="border: none;border-collapse: collapse;width:102pt;">
    <tbody>
        <tr>
            <td style="color:black;font-size:15px;font-weight:700;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;width:51pt;">Problem</td>
            <td style="color:black;font-size:15px;font-weight:700;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-left:none;width:51pt;">Points</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">1</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.5</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">2</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.5</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">3</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">6</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">4</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">1</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">5</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">2</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">6</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">2</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">7.1</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">2</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">7.2</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">2</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">8.1</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">2.5</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">8.2</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">2</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">8.3</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">1</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">8.4</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">2</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">9</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">1.5</td>
        </tr>
    </tbody>
</table>
<p></p>




<hr>

### <u>STAT107 Review</u>
Part of this individual assignment will involve a review of STAT107 code that you have seen before including the following. Click on the links below if you need a refresher.

1. <a href="https://discovery.cs.illinois.edu/learn/Basics-of-Data-Science-with-Python/DataFrames-with-Conditionals/">Row Filtering</a>
2. <a href="https://discovery.cs.illinois.edu/guides/Modifying-DataFrames/dataframe-groupby/">The groupby() and reset_index() Functions</a>
2. <a href="https://discovery.cs.illinois.edu/guides/Modifying-DataFrames/sorting-a-dataframe-with-pandas/">The sort_values() Function</a>


### <u>Tutorial you Will Need</u>: Utilizing the *.shape* attribute

The **.shape** attribute for a given dataframe returns a **tuple** of two values, the first value in the tuple is the number of rows, the second value is the number of columns.

In [2]:
import pandas as pd
demo_df=pd.DataFrame({'a': [1,2,3], 'b': [4,5,6]})
demo_df

a  b
0  1  4
1  2  5
2  3  6

We can index a tuple with 0 and 1 indices to extract the number of rows and columns from the dataframe separately.

In [3]:
#Number of rows and columns
demo_df.shape

(3, 2)

In [4]:
#Number of rows
demo_df.shape[0]

3

In [5]:
#Number of columns
demo_df.shape[1]

2

<hr>

### Imports

### 1. Read the csv
First, read the **steam_sample_400.csv** file into a dataframe and display the first five rows of the dataframe.

In [6]:
df = pd.read_csv('steam_sample_400.csv')
df.head()

user_id    game_name  activity hours_played_if_play
0  218323237       Dota 2  purchase                    1
1  218323237       Dota 2      play                 14.8
2  218323237  Tomb Raider  purchase                    1
3  218323237  Tomb Raider      play                  8.8
4  218323237         TERA  purchase                    1

### 2. How many rows does the dataframe have? (Use Python code to show this).

In [7]:
df.shape[0]


6307

### 3. Data cleaning and data type checking

We plan to analyze the **"hours_played_if_play"** variable in this assignment. In order to ensure that our subsequent functions work appropriately on this numerical variable, you should do the following.
1. Identify any values (if any) that have been encoded in the csv to represent a **missing value**. 
2. Make sure that the values in your **"hours_played_if_play"** variable are represented as numerical values (either float64 or int64 object types). If not, fix this using the methods we discussed in class.
3. PERMANENTLY delete any rows in your dataframe that have missing values (both explicit or implict). Remember, some "missing values", in the way that they're currently encoded in the csv may not be initially detectable by Python. It is your job to figure out what these implicit missing values are and make it so Python does detect them. And then delete any rows that have them.

In [8]:
df = pd.read_csv('steam_sample_400.csv',na_values=['data not found','data missing'])
temp = df.isna()
temp.sum()
pd.set_option('display.max_rows',None)
pd.reset_option('display.max_rows')


In [9]:
df = df.dropna().reset_index()
df

index    user_id                                     game_name  \
0         0  218323237                                        Dota 2   
1         1  218323237                                        Dota 2   
2         2  218323237                                   Tomb Raider   
3         3  218323237                                   Tomb Raider   
4         4  218323237                                          TERA   
...     ...        ...                                           ...   
6293   6302    7519923                                Counter-Strike   
6294   6303    7519923                 Counter-Strike Condition Zero   
6295   6304    7519923  Counter-Strike Condition Zero Deleted Scenes   
6296   6305  103963922                               Team Fortress 2   
6297   6306  103963922                               Team Fortress 2   

      activity  hours_played_if_play  
0     purchase                   1.0  
1         play                  14.8  
2     purchase                   1.0  
3         play                   8.8  
4     purchase                   1.0  
...        ...                   ...  
6293      play                  35.0  
6294  purchase                   1.0  
6295  purchase                   1.0  
6296  purchase                   1.0  
6297      play                   0.4  

[6298 rows x 5 columns]

In [10]:
df.dtypes

index                     int64
user_id                   int64
game_name                object
activity                 object
hours_played_if_play    float64
dtype: object

### 4. Number of Dropped Rows 

How many rows with missing values did you delete from your original dataframe?

In [11]:
temp.shape[0]-df.shape[0]

9

In [12]:
df

index    user_id                                     game_name  \
0         0  218323237                                        Dota 2   
1         1  218323237                                        Dota 2   
2         2  218323237                                   Tomb Raider   
3         3  218323237                                   Tomb Raider   
4         4  218323237                                          TERA   
...     ...        ...                                           ...   
6293   6302    7519923                                Counter-Strike   
6294   6303    7519923                 Counter-Strike Condition Zero   
6295   6304    7519923  Counter-Strike Condition Zero Deleted Scenes   
6296   6305  103963922                               Team Fortress 2   
6297   6306  103963922                               Team Fortress 2   

      activity  hours_played_if_play  
0     purchase                   1.0  
1         play                  14.8  
2     purchase                   1.0  
3         play                   8.8  
4     purchase                   1.0  
...        ...                   ...  
6293      play                  35.0  
6294  purchase                   1.0  
6295  purchase                   1.0  
6296  purchase                   1.0  
6297      play                   0.4  

[6298 rows x 5 columns]

### 5. Separate the dataframes. 

Next, in this analysis we would like to analyze the game purchase information and game play information separately. First create two new dataframes:
* one that is comprised of the purchase rows
* the other that is comprised of the play rows.

In [13]:
df_play = df[df.activity == "play"]
df_play

index    user_id                        game_name activity  \
1         1  218323237                           Dota 2     play   
3         3  218323237                      Tomb Raider     play   
5         5  218323237                             TERA     play   
7         7  126340495                           Dota 2     play   
9         9  126340495  Counter-Strike Global Offensive     play   
...     ...        ...                              ...      ...   
6287   6296   73761548       Sid Meier's Civilization V     play   
6289   6298  138590829                           Dota 2     play   
6291   6300  190460990                           Dota 2     play   
6293   6302    7519923                   Counter-Strike     play   
6297   6306  103963922                  Team Fortress 2     play   

      hours_played_if_play  
1                     14.8  
3                      8.8  
5                      0.9  
7                   1784.0  
9                    626.0  
...                    ...  
6287                  92.0  
6289                1026.0  
6291                   1.5  
6293                  35.0  
6297                   0.4  

[2314 rows x 5 columns]

In [14]:
df_purchase = df[df.activity == 'purchase']
df_purchase

index    user_id                                     game_name  \
0         0  218323237                                        Dota 2   
2         2  218323237                                   Tomb Raider   
4         4  218323237                                          TERA   
6         6  126340495                                        Dota 2   
8         8  126340495               Counter-Strike Global Offensive   
...     ...        ...                                           ...   
6290   6299  190460990                                        Dota 2   
6292   6301    7519923                                Counter-Strike   
6294   6303    7519923                 Counter-Strike Condition Zero   
6295   6304    7519923  Counter-Strike Condition Zero Deleted Scenes   
6296   6305  103963922                               Team Fortress 2   

      activity  hours_played_if_play  
0     purchase                   1.0  
2     purchase                   1.0  
4     purchase                   1.0  
6     purchase                   1.0  
8     purchase                   1.0  
...        ...                   ...  
6290  purchase                   1.0  
6292  purchase                   1.0  
6294  purchase                   1.0  
6295  purchase                   1.0  
6296  purchase                   1.0  

[3984 rows x 5 columns]

### 6. Percentage of Game Purchases that are Actually Played

What percentage of all total game purchases in this dataset are actually played by the player that purchased them?

*You can assume that a game and user id that has a 'play' row ALSO MUST have a 'purchase' row (but not the other way around).*

**For full credit, use automated Python code to find your answer. (Ie. don't put something like '30/50', where you found the 30 and 50 from the output of other code.)**

In [15]:
df_play.shape[0]/df_purchase.shape[0]

0.5808232931726908

### 7. Top Game Purchaser

Next, we'd like to learn more about the people who *purchase* the most amount of games.

#### 7.1. Most Amount of Games Purchased

What is the user id of the player who has purchased the most amount of games in this dataset, and how many games have they purchased?

In [16]:
df.groupby('user_id').agg('count').reset_index().nlargest(1,'activity')

user_id  index  game_name  activity  hours_played_if_play
16  27279261    288        288       288                   288

User id 27279261 has purchased 288 games.

#### 7.2. How many games has this user actually played?

What percentage of the purchased games did this player from 7.1 actually play?

In [17]:
df_play[df_play.user_id == 27279261].shape[0]/df_purchase[df_purchase.user_id == 27279261].shape[0]

0.5567567567567567

## 8. Examining Some Game "All-Star" Players

Next, we'd like to examine the play vs. purchase behavior of a few games' "all star players". We will consider a game's "all star player" to be the one who has played the game the longest in this dataset.

### 8.1. Each Game's Highest Play Time

In a single dataframe below, display the highest amount of time that each game was played.

For full credit, your final resulting dataframe should:
1. Be comprised of *two columns*
 * `game_name`
 * `hours_played_if_play`
2. And should have an index going from 0,1,2,... etc.

In [18]:
df_all_star = df_play[['game_name','hours_played_if_play']].groupby('game_name').agg('sum').reset_index()
df_all_star

game_name  hours_played_if_play
0                                  7 Days to Die                 533.2
1                                        8BitBoy                  24.2
2                          A Valley Without Wind                   1.2
3                        A Valley Without Wind 2                   0.1
4    ACE COMBAT ASSAULT HORIZON Enhanced Edition                   7.7
..                                           ...                   ...
838                          Zombie Panic Source                   0.5
839                                        eden*                   4.3
840                 realMyst Masterpiece Edition                   7.5
841                resident evil 4 / biohazard 4                   3.8
842                                    theHunter                   0.8

[843 rows x 2 columns]

### 8.2. Ranking the Games by Highest Maximum Game Play Time

Sort your final resulting dataframe from 8.1 by the number of hours that the game was played by the all-star player *in descending order*.

Overwrite your dataframe, so it remains sorted in this order.

In [19]:
df_all_star=df_all_star.sort_values('hours_played_if_play',ascending=False)
df_all_star

game_name  hours_played_if_play
225                              Dota 2               51518.2
154     Counter-Strike Global Offensive               11011.9
616          Sid Meier's Civilization V                7469.2
151                      Counter-Strike                6720.4
60    Baldur's Gate II Enhanced Edition                3898.0
..                                  ...                   ...
621             Skullgirls Endless Beta                   0.1
496        One Way To Die Steam Edition                   0.1
510             Passing Pineview Forest                   0.1
57             Back to Dinosaur Island                    0.1
377  I Have No Mouth, and I Must Scream                   0.1

[843 rows x 2 columns]

### 8.3. Top 3 Games with the Most Dedicated "All-Star Player"

Which three games in this dataset had the most "dedicated" "all star players"? Or in other words, which three games were played the longest amount of time by a particular player?

In [23]:
df_all_star.head(3)

game_name  hours_played_if_play
225                           Dota 2               51518.2
154  Counter-Strike Global Offensive               11011.9
616       Sid Meier's Civilization V                7469.2

### 8.4. Identifying these "All Star Players"

For these three games that you selected in 8.3 above, what is the user id of their respective "all star player"?

In [28]:
df_play[(df_play.game_name == "Dota 2")].nlargest(1, "hours_played_if_play")

index   user_id game_name activity  hours_played_if_play
1147   1150  67694595    Dota 2     play                5608.0

In [29]:
df_play[df_play.game_name == "Counter-Strike Global Offensive"].nlargest(1, "hours_played_if_play")

index    user_id                        game_name activity  \
952    955  138115430  Counter-Strike Global Offensive     play   

     hours_played_if_play  
952                1498.0

In [30]:
df_play[df_play.game_name == "Sid Meier's Civilization V"].nlargest(1, "hours_played_if_play")

index    user_id                   game_name activity  \
5963   5972  128065805  Sid Meier's Civilization V     play   

      hours_played_if_play  
5963                5002.0

## 9. Learning about the All-Star Players

Finally, let's learn more about the play vs. purchase behavior of these three "all star" players that we identified in #8. Specifically for each of these three player user ids that you identified in #8, what percentage of games that they bought did they actually play?

In [31]:
df_play[df_play.user_id == 138115430].shape[0]/df_purchase[df_purchase.user_id == 138115430].shape[0]

0.36363636363636365

In [32]:
df_play[df_play.user_id == 67694595].shape[0]/df_purchase[df_purchase.user_id == 67694595].shape[0]

0.631578947368421

In [33]:
df_play[df_play.user_id == 128065805].shape[0]/df_purchase[df_purchase.user_id == 128065805].shape[0]

1.0